tests/test_edgeql_data_migration.py

```diff
@@ -0,0 +1,2526 @@
+
+        # Fix the data.
+        await self.con.execute(r"""
+            UPDATE Base
+            SET {
+                foo := 'base_09',
+            };
+        """)
+
+        # Migrate to same state as before now that the data is fixed.
+        await self._migrate(new_state)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    foo,
+                };
+            """,
+            [{
+                'foo': 'base_09',
+            }],
+        )
+
+    async def test_edgeql_migration_11(self):
+        await self.con.execute("""
+            SET MODULE test;
+        """)
+        await self._migrate(r"""
+            type Base {
+                property foo -> str;
+            }
+        """)
+        await self.con.execute(r"""
+            INSERT Base {
+                foo := 'base_11',
+            };
+        """)
+
+        await self._migrate(r"""
+            type Child;
+
+            type Base {
+                # change property to link with same name
+                link foo -> Child {
+                    # add a constraint
+                    constraint exclusive;
+                }
+            }
+        """)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    foo,
+                };
+            """,
+            [{
+                'foo': None,
+            }],
+        )
+
+    async def test_edgeql_migration_12(self):
+        await self.con.execute("""
+            SET MODULE test;
+        """)
+        await self._migrate(r"""
+            type Child;
+
+            type Base {
+                property foo -> str {
+                    constraint exclusive;
+                }
+
+                link bar -> Child {
+                    constraint exclusive;
+                }
+            }
+        """)
+        data = await self.con.fetchall(r"""
+            SELECT (
+                INSERT Base {
+                    foo := 'base_12',
+                    bar := (INSERT Child)
+                })
+            {
+                foo,
+                bar: {id}
+            };
+        """)
+
+        await self._migrate(r"""
+            type Child;
+
+            type Base {
+                # drop constraints
+                property foo -> str;
+                link bar -> Child;
+            }
+        """)
+
+        await self.assert_query_result(
+            r"""
+                SELECT Base {
+                    foo,
+                    bar: {id}
+                };
+            """,
+            [{
+                'foo': 'base_12',
+                'bar': {'id': data[0].bar.id}
+            }],
+        )
+
+    @test.xfail('''
+        edgedb.errors.InternalServerError: column Derived~2.bar does
+        not exist
+
+        A problem arises when a link was inherited, but then needs
+        to be "moved up" in the inheritance hierarchy. This is the
```

Comment:
I usually visualize the inheritance hierarchy like an org chart, i.e the most generic type on top, so "moved up" here is confusing.

Reply:
I changed the wording to refer to derived types instead.